File: Chapter06/Problem06F/Main.java
import java.io.*;
import java.lang.*;
import java.util.*;


public class Main {
	public static final Scanner scanner = new Scanner(System.in);
	public static final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

	public static ArrayList<Bomb> getRemovableOrders(int n, Bomb[] bombs) {
		ArrayList<Bomb> removedList = new ArrayList<>();

		// poll시에 제거 가능한 폭탄들 중 가장 인덱스가 작은 번호가 반환될 수 있도록
		// PriorityQueue를 사용한다. 폭탄들의 우선순의는 compareTo 메소드로 정의했다.
		PriorityQueue<Bomb> removableBombs = new PriorityQueue<>();

		for (int i = 0; i < n; i += 1) {
			if (bombs[i].getChildCount() == 0) {
				removableBombs.add(bombs[i]);
			}
		}

		while (removableBombs.isEmpty() == false) {
			Bomb b = removableBombs.poll();
			b.remove();

			removedList.add(b);

			ArrayList<Bomb> parents = b.getParentBombs();
			for (int i = 0; i < parents.size(); i += 1) {
				Bomb p = parents.get(i);
				if (p.getChildCount() == 0) {
					removableBombs.add(p);
				}
			}
		}

		// 모든 폭탄이 제거되지 않았다면 null을 반환한다.
		if (removedList.size() != n) {
			return null;
		}
		return removedList;
	}

	public static void testCase(int caseIndex) throws Exception {
		int n = scanner.nextInt();
		int m = scanner.nextInt();

		Bomb[] bombs = new Bomb[n];
		for (int i = 0; i < n; i += 1) {
			bombs[i] = new Bomb(i);
		}

		for (int i = 0; i < m; i++) {
			int u = scanner.nextInt() - 1;
			int v = scanner.nextInt() - 1;
			Bomb parent = bombs[u];
			Bomb child = bombs[v];

			child.addParentBombs(parent);
		}

		ArrayList<Bomb> orders = getRemovableOrders(n, bombs);

		if (orders == null) {
			writer.write("NO\n");
		} else {
			for (int i = 0; i < n; i++) {
				Bomb b = orders.get(i);

				if (i > 0) {
					writer.write(" ");
				}
				writer.write(String.valueOf(b.index + 1));
			}
			writer.write("\n");
		}
	}

	public static void main(String[] args) throws Exception {
		int caseSize = scanner.nextInt();

		for (int caseIndex = 1; caseIndex <= caseSize; caseIndex += 1) {
			testCase(caseIndex);
		}

		writer.flush();
		writer.close();
	}

}

class Bomb implements Comparable<Bomb> {
	public final int index;

	private int childCount;
	private ArrayList<Bomb> parentBombs;

	Bomb(int index) {
		this.index = index;
		this.parentBombs = new ArrayList<>();
		this.childCount = 0;
	}

	public void addParentBombs(Bomb parentBomb) {
		this.parentBombs.add(parentBomb);
		parentBomb.childCount += 1;
	}

	public ArrayList<Bomb> getParentBombs() {
		return this.parentBombs;
	}

	public int getChildCount() {
		return childCount;
	}

	public void remove() {
		for (int i = 0; i < parentBombs.size(); i += 1) {
			Bomb p = parentBombs.get(i);
			p.childCount -= 1;
		}
	}

	@Override
	public int compareTo(Bomb o) {
		// 인덱스가 작을수록 priorityQueue에서 먼저 poll 되도록
		// 우선순위를 정의한다.
		return this.index - o.index;
	}
}

File: Chapter06/Problem06F/main.cpp
#include<bits/stdc++.h>
using namespace std;

class Bomb {
public:
	int index;
	int childCnt;
	vector<Bomb*>parentBombs;

	Bomb() {
		childCnt = 0;
	};
	Bomb(int _index) {
		index = _index;
		childCnt = 0;
	}

	void addParentBombs(Bomb* parentBomb) {
		parentBombs.push_back(parentBomb);
		parentBomb->childCnt++;
	}
};

class Cmp {
public:
	bool operator()(Bomb *a, Bomb *b) {
		return a->index > b->index;
	}
};

vector<Bomb*> isAllRemoveable(int n, vector<Bomb*> &bombs) {

	//제거된 폭탄 목록
	vector<Bomb*>removedList;

	//제거 가능한 폭탄 목록
	//priority queue를 사용한다. 폭탄들의 우선순위는 Cmp class로 정의했다.
	priority_queue<Bomb*, vector<Bomb*>, Cmp>removableBombs;

	// 현재 연쇄 폭탄이 없는 폭탄들은 제거 가능한 목록에 추가한다.
	for (int i = 0; i < n; i++) {
		if (bombs[i]->childCnt == 0) {
			removableBombs.push(bombs[i]);
		}
	}

	// 제거 가능한 폭탄이 남아있는 동안 반복한다.
	while (removableBombs.size()) {

		// 현재 연쇄폭탄이 없는 폭탄 b를 제거하고, 제거된 폭탄 갯수를 증가시킨다.
		Bomb *b = removableBombs.top();
		removableBombs.pop();
		removedList.push_back(b);

		//제거된 폭탄의 부모 폭탄들의 child갯수를 1씩 내린다.
		//만약 child갯수가 0이 된 폭탄이 있으면 제거 가능한 폭탄 목록에 추가한다.
		vector<Bomb*> &parents = b->parentBombs;
		for (int i = 0; i < parents.size(); i++) {
			Bomb *p = parents[i];
			p->childCnt--;
			if (p->childCnt == 0) {
				removableBombs.push(p);
			}
		}
	}

	//제거한 폭탄들 목록은 return한다.
	return removedList;
}

void testCase() {
	int n, m;
	scanf("%d %d", &n, &m);

	vector<Bomb*>bombs(n);

	for (int i = 0; i < n; i++) {
		bombs[i] = new Bomb(i + 1);
	}

	for (int i = 0; i < m; i++) {
		int u, v;
		scanf("%d %d", &u, &v);
		u -= 1;
		v -= 1;
		Bomb* parent = bombs[u];
		Bomb* child = bombs[v];
		child->addParentBombs(parent);
	}
	vector<Bomb*> removable = isAllRemoveable(n, bombs);
	if (removable.size() == n) {
		for (int i = 0; i < removable.size(); i++) {
			if (i)printf(" ");
			printf("%d", removable[i]->index);
		}
	}
	else {
		printf("NO");
	}
	printf("\n");

	for (int i = 0; i < n; i++) {
		delete bombs[i];
	}
}

int main() {
	int t; scanf("%d", &t);
	while (t--) {
		testCase();
	}
	return 0;
}
